### ISSUE TITLE
SMT cannot be used with examples/pe 

### ISSUE THREAD

**Issue by rdaly525** [cdonovick/peak#100]
Currently we have the following in examples/pe/sim.py

```
 _, aluinst = inst.alu.match()
data = [self.data[i](aluinst.data_modes[i],aluinst.data_consts[i], data[i],clk_en) for i in range(num_inputs)]
alu_res, alu_res_p, C, V = aluinst.eval(*(data + [bit0]))
```

where 'inst' is an instance of the Sum type Sum[A,B,C,D,E] and where each of A,B,C,D,E have its own implementation of the method 'eval'

I know this is a very clean generic way of writing the ALU (with getting common fields and using the specialized method 'eval'), but this fundamentally does not work if you have an SMT free variable representing the instance of the Sum type (Inst). This is because the only way to get the Sum subtype is if the instance of Sum has a constant value for it's sum tag. This is not going to be true when we are using SMT for rewrite rule discovery. 

I want to open this issue to discuss possible solutions that are still clean and generic, but allows us to evaluate on an SMT free variable.

**Comment by phanrahan** [comment]
That was an experiment based on code that Nikhyl had written. He wanted to combine the isa and the semantics, ala TIE. I suggest we rewrite it ala the way we are doing it in the other examples.

**Comment by rdaly525** [comment]
I agree that to get stuff merged, we should rewrite it (or disable it for now), but I do think it is an interesting design pattern.

There might be a way to do the function specialization with a magic "peak_abstract_method" decorator that evaluates the functions for all the subtypes and muxes on them via the sum tag.
